File: sections/66_master_math_mixed_contest/problems/e_black_box_walks/solution.cpp
#include <bits/stdc++.h>
using namespace std;

using int64 = long long;
constexpr int64 MOD = 998244353;

int64 power(int64 base, int64 exponent) {
    int64 result = 1;
    while (exponent) {
        if (exponent & 1) result = result * base % MOD;
        base = base * base % MOD;
        exponent >>= 1;
    }
    return result;
}

vector<int64> berlekamp_massey(const vector<int64> &sequence) {
    vector<int64> current{1}, saved{1};
    int length = 0, shift = 1;
    int64 saved_discrepancy = 1;
    for (int position = 0; position < static_cast<int>(sequence.size()); ++position) {
        int64 discrepancy = sequence[position];
        for (int i = 1; i <= length; ++i)
            discrepancy = (discrepancy + current[i] * sequence[position - i]) % MOD;
        if (discrepancy == 0) {
            ++shift;
            continue;
        }
        vector<int64> previous = current;
        int64 scale = discrepancy * power(saved_discrepancy, MOD - 2) % MOD;
        if (current.size() < saved.size() + shift)
            current.resize(saved.size() + shift);
        for (int i = 0; i < static_cast<int>(saved.size()); ++i) {
            current[i + shift] = (current[i + shift] - scale * saved[i]) % MOD;
            if (current[i + shift] < 0) current[i + shift] += MOD;
        }
        if (2 * length <= position) {
            length = position + 1 - length;
            saved = move(previous);
            saved_discrepancy = discrepancy;
            shift = 1;
        } else {
            ++shift;
        }
    }
    vector<int64> recurrence(length);
    for (int i = 0; i < length; ++i)
        recurrence[i] = (MOD - current[i + 1]) % MOD;
    return recurrence;
}

vector<int64> combine(const vector<int64> &a, const vector<int64> &b,
                      const vector<int64> &recurrence) {
    int order = recurrence.size();
    vector<int64> product(2 * order - 1);
    for (int i = 0; i < order; ++i)
        for (int j = 0; j < order; ++j)
            product[i + j] = (product[i + j] + a[i] * b[j]) % MOD;
    for (int degree = 2 * order - 2; degree >= order; --degree)
        for (int back = 1; back <= order; ++back)
            product[degree - back] = (
                product[degree - back] + product[degree] * recurrence[back - 1]
            ) % MOD;
    product.resize(order);
    return product;
}

int64 nth_term(const vector<int64> &sequence, const vector<int64> &recurrence,
               unsigned long long index) {
    if (index < sequence.size()) return sequence[index];
    int order = recurrence.size();
    if (order == 0) return 0;
    vector<int64> result(order), base(order);
    result[0] = 1;
    if (order == 1) base[0] = recurrence[0];
    else base[1] = 1;
    while (index) {
        if (index & 1) result = combine(result, base, recurrence);
        base = combine(base, base, recurrence);
        index >>= 1;
    }
    int64 answer = 0;
    for (int i = 0; i < order; ++i)
        answer = (answer + result[i] * sequence[i]) % MOD;
    return answer;
}

int main() {
    ios::sync_with_stdio(false);
    cin.tie(nullptr);
    int n, m, start, target;
    unsigned long long length;
    if (!(cin >> n >> m >> length >> start >> target)) return 0;
    --start; --target;
    struct Edge { int from, to; int64 weight; };
    vector<Edge> edges(m);
    for (auto &[from, to, weight] : edges) {
        cin >> from >> to >> weight;
        --from; --to;
    }

    vector<int64> state(n), sequence(2 * n + 1);
    state[start] = 1;
    for (int step = 0; step <= 2 * n; ++step) {
        sequence[step] = state[target];
        vector<int64> next(n);
        for (auto [from, to, weight] : edges)
            next[to] = (next[to] + state[from] * weight) % MOD;
        state.swap(next);
    }
    vector<int64> recurrence = berlekamp_massey(sequence);
    cout << nth_term(sequence, recurrence, length) << '\n';
}

File: sections/66_master_math_mixed_contest/problems/e_black_box_walks/solution.py
import sys

MOD = 998_244_353


def berlekamp_massey(sequence: list[int]) -> list[int]:
    current, saved = [1], [1]
    length, shift, saved_discrepancy = 0, 1, 1
    for position, term in enumerate(sequence):
        discrepancy = term
        for i in range(1, length + 1):
            discrepancy = (discrepancy + current[i] * sequence[position - i]) % MOD
        if discrepancy == 0:
            shift += 1
            continue
        previous = current[:]
        scale = discrepancy * pow(saved_discrepancy, MOD - 2, MOD) % MOD
        current += [0] * max(0, len(saved) + shift - len(current))
        for i, value in enumerate(saved):
            current[i + shift] = (current[i + shift] - scale * value) % MOD
        if 2 * length <= position:
            length = position + 1 - length
            saved = previous
            saved_discrepancy = discrepancy
            shift = 1
        else:
            shift += 1
    return [(-current[i]) % MOD for i in range(1, length + 1)]


def combine(left: list[int], right: list[int], recurrence: list[int]) -> list[int]:
    order = len(recurrence)
    product = [0] * (2 * order - 1)
    for i, a in enumerate(left):
        for j, b in enumerate(right):
            product[i + j] = (product[i + j] + a * b) % MOD
    for degree in range(2 * order - 2, order - 1, -1):
        value = product[degree]
        for back, coefficient in enumerate(recurrence, 1):
            product[degree - back] = (
                product[degree - back] + value * coefficient
            ) % MOD
    return product[:order]


def nth_term(sequence: list[int], recurrence: list[int], index: int) -> int:
    if index < len(sequence):
        return sequence[index]
    order = len(recurrence)
    if order == 0:
        return 0
    result = [1] + [0] * (order - 1)
    base = [recurrence[0]] if order == 1 else [0, 1] + [0] * (order - 2)
    while index:
        if index & 1:
            result = combine(result, base, recurrence)
        base = combine(base, base, recurrence)
        index >>= 1
    return sum(result[i] * sequence[i] for i in range(order)) % MOD


def main() -> None:
    data = list(map(int, sys.stdin.buffer.read().split()))
    if not data:
        return
    n, m, walk_length, start, target = data[:5]
    start -= 1
    target -= 1
    edges = []
    index = 5
    for _ in range(m):
        u, v, weight = data[index : index + 3]
        index += 3
        edges.append((u - 1, v - 1, weight))
    state = [0] * n
    state[start] = 1
    sequence = []
    for _ in range(2 * n + 1):
        sequence.append(state[target])
        next_state = [0] * n
        for u, v, weight in edges:
            next_state[v] = (next_state[v] + state[u] * weight) % MOD
        state = next_state
    recurrence = berlekamp_massey(sequence)
    print(nth_term(sequence, recurrence, walk_length))


if __name__ == "__main__":
    main()
